### PR TITLE
Attempt To Clean Up Bracket Matcher Accommodation

### PR DESCRIPTION
**Pull request by joefitzgerald**
- Also allow space to trigger autocomplete, disabled by default
- The size of the newText / oldText comparison logic feels dirty

@benogle this might still feel like a hack, but it doesn't create two garbage arrays on each keystroke.

### CODE REVIEW

File: lib/main.coffee [hunk 1]
```diff
@@ -86,19 +86,25 @@ module.exports =
       type: 'boolean'
       default: true
       order: 13
+    spaceTriggersAutocomplete:
```
Comment (benogle) [inline lib/main.coffee]: I dont think there should be an option for it. I think the provider should decide. For example an objective c provider might want to trigger after `[self |`, but the fuzzy provider should never trigger after space.

File: lib/autocomplete-manager.coffee [hunk 2]
```diff
@@ -356,7 +359,12 @@ class AutocompleteManager
     return if @disposed
     return @hideSuggestionList() if @compositionInProgress
     autoActivationEnabled = atom.config.get('autocomplete-plus.enableAutoActivation')
-    wouldAutoActivate = newText.trim().length in [1..2] or ((@backspaceTriggersAutocomplete or @suggestionList.isActive()) and oldText.trim().length in [1..2])
+    # TODO: Can we avoid bracket-matcher causing a 2-character change in the buffer changed event (can we get each character separately)?
+    wouldAutoActivate = false
+    if @spaceTriggersAutocomplete
+      wouldAutoActivate = (newText.length is 1 or newText in @bracketMatcherPairs) or ((@backspaceTriggersAutocomplete or @suggestionList.isActive()) and (oldText.trim().length is 1 or oldText in @bracketMatcherPairs))
```
Comment (benogle) [inline lib/autocomplete-manager.coffee]: We should just do this all the time. Providers can trim the prefix and check the length if they dont support space.

File: appveyor.yml [hunk 3]
```diff
@@ -3,15 +3,14 @@ version: "{build}"
 os: Windows Server 2012 R2
 
 install:
-  - appveyor DownloadFile https://atom.io/download/windows -FileName AtomSetup.exe
```
Comment (benogle) [inline appveyor.yml]: Did you mean to commit the appveyor stuff here?

**Comment by benogle** [comment]
This is better as far as the bracket matcher stuff is concerned.

File: lib/main.coffee (hunk 1 above)
Comment (joefitzgerald) [inline lib/main.coffee]: Providers aren't available at this step in the process. What you're saying is that we should always trigger on space, but then let provider(s) determine if they should respond in a 'whitespace-only' scenario.

File: appveyor.yml (hunk 3 above)
Comment (joefitzgerald) [inline appveyor.yml]: Yeah, I wanted the stupid Windows build to stop failing. Turns out it's the package installer that's flakey, and by association the Chocolatey install also (i.e. it didn't really help).

File: lib/autocomplete-manager.coffee (hunk 2 above)
Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: Do you think I should remove the trim() in the oldText also? Additionally, should we do some inspection for \t, \r, \n and prevent activation in those scenarios?

File: lib/main.coffee (hunk 1 above)
Comment (benogle) [inline lib/main.coffee]: > we should always trigger on space, but then let provider(s) determine if they should respond in a 'whitespace-only' scenario.

yes

File: lib/autocomplete-manager.coffee (hunk 2 above)
Comment (benogle) [inline lib/autocomplete-manager.coffee]: > should we do some inspection for \t, \r, \n and prevent activation in those scenarios?

That seems like a good check. Maybe whitelist approach, so if `newText is ' ' or newText.trim().length is 1`

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: I like that idea. Whitelist approach implemented.

**Comment by joefitzgerald** [comment]
So apparently a bunch of our tests suck, because they break now that we're allowing space to trigger.

File: lib/autocomplete-manager.coffee [hunk 4]
```diff
@@ -356,7 +357,8 @@ class AutocompleteManager
     return if @disposed
     return @hideSuggestionList() if @compositionInProgress
     autoActivationEnabled = atom.config.get('autocomplete-plus.enableAutoActivation')
-    wouldAutoActivate = newText.trim().length in [1..2] or ((@backspaceTriggersAutocomplete or @suggestionList.isActive()) and oldText.trim().length in [1..2])
+    # TODO: Can we avoid bracket-matcher causing a 2-character change in the buffer changed event (can we get each character separately)?
+    wouldAutoActivate = (newText is ' ' or newText.trim().length is 1 or newText in @bracketMatcherPairs) or ((@backspaceTriggersAutocomplete or @suggestionList.isActive()) and (oldText is ' ' or oldText.trim().length is 1 or oldText in @bracketMatcherPairs))
```
Comment (benogle) [inline lib/autocomplete-manager.coffee]: Can we make this logic more clear somehow? Maybe split it into multiple lines? I'm not sure, but it's a lot to digest in one line.

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: It is; I was assuming you wanted to avoid allocations during this check.

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: See 164c4ab9f4f06cf2e9016206d418cde490462e24 for a clearer version with no additional allocation.

**Comment by joefitzgerald** [comment]
@benogle :thumbsup: / :thumbsdown:?

**Comment by benogle** [comment]
:+1: Does rebasing remove the appveyor file from this changeset? It doesnt really belong in here.

**Comment by joefitzgerald** [comment]
Closing in deference to #324.

**Pull request closed without merging**